src/layers/msgpass.jl

```diff
@@ -1,7 +1,70 @@
-abstract type MessagePassing <: GraphNet end
+# Adapted message passing from paper 
+# "Relational inductive biases, deep learning, and graph networks"
+"""
+    MessagePassing
+
+The abstract type from which all message passing layers are derived.
+
+Related methods are [`propagate`](@ref), [`message`](@ref),
+[`update`](@ref), [`update_edge`](@ref), and [`update_global`](@ref). 
+"""
+abstract type MessagePassing end
+
+"""
+    propagate(mp::MessagePassing, fg::FeaturedGraph, aggr)
+    propagate(mp::MessagePassing, fg::FeaturedGraph, E, X, u, aggr)
+
+Perform the sequence of operation implementing the message-passing scheme
+and updating node, edge, and global features `X`, `E`, and `u` respectively.
+
+The computation involved is the following:
+
+```julia
+M = compute_batch_message(mp, fg, E, X, u) 
+E = update_edge(mp, M, E, u)
+M̄ = aggregate_neighbors(mp, aggr, fg, M)
+X = update(mp, M̄, X, u)
+u = update_global(mp, E, X, u)
+```
```

Comment:
This is really annoying. It confuses the definition of message-passing network and graph network. We shouldn't break the definitions. By definition from [paper](https://arxiv.org/pdf/1704.01212.pdf),  there are two functions to be defined:

- `message`: it process information from node feature of itself and it's neighbors, as well as edge feature, and then gives a message as output.
- `update`: Message after aggregation is passed to `update` function. It takes message and node feature itself as input and output a new node feature for next layer.

Between these, an aggregation function is needed to apply to aggregate messages.

But for graph network is not the case. Graph network is a more general scheme that include message-passing network. [Paper](https://arxiv.org/pdf/1806.01261.pdf) defines a series of APIs and the algorithm for graph network. It contains:

- Update function:
  - For node
  - For edge
  - For global
- Aggregate function:
  - For aggregating edge feature into node feature
  - For aggregating edge feature into global feature
  - For aggregating node feature into global feature

This architecture is flexible because the functions above should be optional. They can be "turn on/off". If the definition is given by user, thus, it is "turned on". Thus, user can combine their own neural network by giving definitions or not.

Thus, message-passing network is one of the specialization of graph network. It is specialize by following:

- `message` is the update function of edge in graph network.
- aggregation in message-passing is the aggregation from edge feature into node feature in graph network.
- `update` is the update function of node in graph network.

I think engineering must be done by satisfying the definitions above. If the behavior is different, we shouldn't call it the same name. `MessagePassing` is for message-passing network and `GraphNet` is for graph network.